src/Cardano/CoinSelection/Random.hs

```diff
@@ -118,55 +220,72 @@ payForOutputs
     -> NonEmpty TxOut
     -> UTxO
     -> ExceptT (ErrCoinSelection e) m (CoinSelection, UTxO)
-payForOutputs opt outs utxo = do
-    let descending = NE.toList . NE.sortBy (flip $ comparing coin)
-    let nOuts = fromIntegral $ NE.length outs
-    let maxN = fromIntegral $ maximumInputCount opt nOuts
-    randomMaybe <- lift $ runMaybeT $
-        foldM makeSelection (maxN, utxo, []) (descending outs)
-    case randomMaybe of
-        Just (maxN', utxo', res) -> do
-            (_, sel, remUtxo) <- lift $
-                foldM improveTxOut (maxN', mempty, utxo') (reverse res)
-            guard sel $> (sel, remUtxo)
+payForOutputs options outputsRequested utxo = do
+    mRandomSelections <- lift $ runMaybeT $ foldM makeRandomSelection
+        (inputCountMax, utxo, []) outputsDescending
+    case mRandomSelections of
+        Just (inputCountRemaining, utxoRemaining, randomSelections) -> do
+            (_, finalSelection, utxoRemaining') <- lift $ foldM
+                improveSelection
+                    (inputCountRemaining, mempty, utxoRemaining)
+                    (reverse randomSelections)
+            validateSelection finalSelection $>
+                (finalSelection, utxoRemaining')
         Nothing ->
-            selectCoins largestFirst opt outs utxo
+            -- In the case that we fail to generate a selection, delegate to
```

Comment:
once again, if falls back is better expression here maybe we can improve here also;-)

Reply:
Fair enough!

Fixed in 60258d5fbc5bc272b41d8f68df358a5ad75c47da